src/main/java/xfacthd/framedblocks/common/compat/jei/camo/CamoRecipeManagerPlugin.java

```diff
@@ -0,0 +1,239 @@
+        }
+        // calculating the correct outputs here is impossible, leave them to be generated by
+        // CamoCraftingRecipeExtension.onDisplayedIngredientsUpdate
+        return createRecipe(
+                Either.left(framedBlock),
+                camoOne,
+                camoTwo,
+                List.of()
+        );
+    }
+
+    private RecipeHolder<CraftingRecipe> createRecipe(
+            Either<ItemStack, TagKey<Item>> frame,
+            Either<ItemStack, TagKey<Item>> camoOne,
+            Either<ItemStack, TagKey<Item>> camoTwo,
+            List<ItemStack> results
+    )
+    {
+        Ingredient frameIngredient = frame.map(Ingredient::of, Ingredient::of);
+        Ingredient copyTool = camoCraftingHelper.getCopyToolIngredient();
+        Ingredient camoOneIngredient = camoOne.map(Ingredient::of, Ingredient::of);
+        Ingredient secondInputStacks = camoTwo.map(Ingredient::of, Ingredient::of);
+        JeiCamoApplicationRecipe recipe = new JeiCamoApplicationRecipe(frameIngredient, copyTool, camoOneIngredient, secondInputStacks, results);
+
+        ResourceLocation resourceLocation = generateId(frame, camoOne, camoTwo);
+        return new RecipeHolder<>(resourceLocation, recipe);
+    }
+
+    private static ResourceLocation generateId(Either<ItemStack, TagKey<Item>> frame, Either<ItemStack, TagKey<Item>> camoOne, Either<ItemStack, TagKey<Item>> camoTwo)
+    {
+        String frameId = mapStackOrTag(frame, List.of(
+                Pair.of(JeiCamoTags.getAllFramesTag(), "all"),
+                Pair.of(JeiCamoTags.getDoubleFramesTag(), "all_double")
+        ));
+        String camoOneId = mapStackOrTag(camoOne, List.of(
+                Pair.of(JeiCamoTags.getCamoBlockExamplesTag(), "examples")
+        ));
+        String camoTwoId = mapStackOrTag(camoTwo, List.of(
+                Pair.of(JeiCamoTags.getCamoBlockExamplesTag(), "examples")
+        ));
+        return Utils.rl("camo_application/jei_generated/" + frameId + "/" + camoOneId + "/" + camoTwoId);
+    }
+
+    private static String mapStackOrTag(Either<ItemStack, TagKey<Item>> value, List<Pair<TagKey<Item>, String>> converters)
+    {
+        return value.map(
+                CamoRecipeManagerPlugin::stackToString,
+                tag -> tagToString(tag, converters)
+        );
+    }
+
+    private static String stackToString(ItemStack itemStack)
+    {
+        if (itemStack.isEmpty())
+        {
+            return "empty";
+        }
+        return BuiltInRegistries.ITEM.getKey(itemStack.getItem()).toLanguageKey();
+    }
+
+    private static String tagToString(TagKey<Item> tag, List<Pair<TagKey<Item>, String>> converters)
+    {
+        for (Pair<TagKey<Item>, String> conv : converters)
+        {
+            if (conv.getFirst().equals(tag))
+            {
+                return conv.getSecond();
+            }
+        }
+        return Utils.sanitizePath(tag.location().toString());
```

Comment:
Please use the language key helper here as well
```suggestion
        return tag.location().toLanguageKey();
```